### PR TITLE
Adding separate tasks to carry out device monitoring for each device type 

### PR DESCRIPTION
**Pull request by madhawap**

### CODE REVIEW

File: features/device-mgt/org.wso2.carbon.device.mgt.server.feature/src/main/resources/conf/cdm-config.xml [hunk 1]
```diff
@@ -48,10 +48,10 @@
             <Platform>windows</Platform>
         </Platforms>
     </PolicyConfiguration>
-    <TaskConfiguration>
-        <Enable>true</Enable>
-        <Frequency>600000</Frequency>
-    </TaskConfiguration>
+    <!--<TaskConfiguration>-->
```
Comment (ruwany) [inline features/device-mgt/org.wso2.carbon.device.mgt.server.feature/src/main/resources/conf/cdm-config.xml:51]: remove if not needed.

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/task/impl/DeviceTaskManagerServiceImpl.java [hunk 2]
```diff
@@ -62,23 +60,27 @@ public void startTask() throws DeviceMgtTaskException {
 
             if (log.isDebugEnabled()) {
                 log.debug("Device details retrieving task is started for the tenant id " + tenantId);
-                log.debug("Device details retrieving task is at frequency of : " + deviceTaskManager.getTaskFrequency());
+//                log.debug("Device details retrieving task is at frequency of : " + deviceTaskManager
```
Comment (ruwany) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/task/impl/DeviceTaskManagerServiceImpl.java:63]: why the empty space?

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/task/impl/DeviceTaskManagerImpl.java [hunk 3]
```diff
@@ -111,26 +95,22 @@ public void addOperations() throws DeviceMgtTaskException {
...
-                } else {
-                    if (log.isDebugEnabled()) {
-                        log.debug("No devices are available to perform the operations.");
-                    }
-                }
 
+            operations = this.getValidOperationNames(); //list operations for each device type
+            devices = deviceManagementProviderService.getAllDevices(deviceType);//list devices for each type
+            if (!devices.isEmpty()) {
+                for (String str : operations) {
```
Comment (ruwany) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/task/impl/DeviceTaskManagerImpl.java:102]: shouldn't we have a null check?

Comment (madhawap) [reply components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/task/impl/DeviceTaskManagerImpl.java:102]: This is fixed with PR https://github.com/wso2/carbon-device-mgt/pull/769

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/task/impl/DeviceTaskManagerServiceImpl.java (hunk 2 above)
Comment (madhawap) [reply components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/task/impl/DeviceTaskManagerServiceImpl.java:63]: This is fixed with https://github.com/wso2/carbon-device-mgt/pull/769

File: features/device-mgt/org.wso2.carbon.device.mgt.server.feature/src/main/resources/conf/cdm-config.xml (hunk 1 above)
Comment (madhawap) [reply features/device-mgt/org.wso2.carbon.device.mgt.server.feature/src/main/resources/conf/cdm-config.xml:51]: This is fixed with https://github.com/wso2/carbon-device-mgt/pull/769

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/internal/DeviceTaskManagerServiceComponent.java [hunk 4]
```diff
@@ -45,20 +49,23 @@
...
             }
             // This will start the device details retrieving task.
-            boolean taskEnable =
-                    DeviceConfigurationManager.getInstance().getDeviceManagementConfig().getTaskConfiguration().
-                            isEnabled();
-            if (taskEnable) {
-                DeviceTaskManagerService taskManagerService = new DeviceTaskManagerServiceImpl();
-                taskManagerService.startTask();
-            }
+//            DeviceTaskManagerService deviceTaskManagerService = new DeviceTaskManagerServiceImpl();
```
Comment (dakshika) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/internal/DeviceTaskManagerServiceComponent.java:61]: it's better to remove his commented code

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/operation/mgt/OperationManagerImpl.java [hunk 5]
```diff
@@ -1034,29 +1038,31 @@ private boolean resetAttemptCount(int enrolmentId) throws OperationManagementExc
...
-                }
+        for(MonitoringOperation op : monitoringOperations){
+            if (operation.getCode().equals(op.getTaskName())) {
+                return true;
             }
-
-
         }
 
+//        for(String dti : taskOperation){
```
Comment (dakshika) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/operation/mgt/OperationManagerImpl.java:1052]: it's better to remove this commented code

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java [hunk 6]
```diff
@@ -1219,24 +1220,37 @@ public int getActivityCountUpdatedAfter(long timestamp) throws OperationManageme
...
-                deviceType = dms.getType();
-                deviceTypeSpecificTasks.put(deviceType, taskOperations);
-            }
-        }
-        return deviceTypeSpecificTasks;
+//        Map<DeviceTypeIdentifier, DeviceManagementService> deviceManagementServiceMap =
+//                pluginRepository.getAllDeviceManagementServices(tenantId);
+        DeviceManagementService dms = pluginRepository.getDeviceManagementService(deviceType, tenantId);
+      // ;
+       // OperationMonitoringTaskConfig operationMonitoringTaskConfig;
```
Comment (dakshika) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java:1229]: it's better to remove this commented code

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/task/impl/DeviceTaskManagerImpl.java [hunk 7]
```diff
@@ -42,65 +40,51 @@
...
 
-    @Override
-    public String getTaskImplementedClazz() throws DeviceMgtTaskException {
-        return DeviceConfigurationManager.getInstance().getDeviceManagementConfig().getTaskConfiguration().
-                getTaskClazz();
     }
 
+//    @Override
+//    public String getTaskImplementedClazz() throws DeviceMgtTaskException {
+//        return DeviceConfigurationManager.getInstance().getDeviceManagementConfig().getTaskConfiguration().
```
Comment (dakshika) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/task/impl/DeviceTaskManagerImpl.java:77]: it's better to remove this commented code

**Pull request merged**